### DISCUSSION TITLE
Can I delete items in dag_run, task_instance and job table?

### DISCUSSION THREAD

**Discussion by pretendhigh** [apache/airflow#19068] [General]
Hello, everyone.

I'am looking for information about the dag_run, task_instance and job tables in the Airflow database schema.  I'd like to clean some old data in these tables.

My question is:
    1. Could we truncate these tables without any risk related to Airflow features ?

I made some test in my local setup (truncate  table), I know when I truncate table dag_run, some dag will rerun if the latest dag_run records are deleted. I think it may be becouse of max(start_date, last_scheduler + 1 interval). For task_instance and job tables, this is not a problem.

So, I decide to delete items in dag_run, task_instance and job tables before 30 days. And of course, I will set "start_date" in my dag file to "days_ago(1)" to avoid dag rerun.
I prefer to have a confirmation from the community to be sure.

**Comment by metov** [comment]
>I know when I truncate table dag_run, some dag will rerun if the latest dag_run records are deleted

If you delete the latest dag run, I would expect that Airflow will think that it still needs to run the DAG, and queue up a new run. This is the same as what would happen if you had the DAG turned off for so long that it missed a run, and then suddenly enabled it. This is a consequence of the scheduler constantly checking "Was I supposed to run this DAG by now?" for every DAG every few minutes.

I suspect that if you delete the run before the last, Airflow will probably not trigger a run (unless you have the backfills enabled).

If you delete task instance rows, I think Airflow will not care because the parent DAG will still be marked as "done" or "failed". When the scheduler sees a DAG that is in these "finished" states, it will not descend into it to check if the tasks are actually complete, even if technically some tasks are still not run. You can test this yourself by doing "mark as success" on a DAG which is still running. The current task will finish, but child tasks will remain something like `no_status`. So I think deleting task instances should be okay so long as it's a reasonably old DAG (ie. don't delete tasks from a DAG that's currently running, or Airflow will think it needs to run that task again).

No idea about the jobs. But I would guess if they are old, completed, and not referenced by anything (because you deleted the relevant tasks etc.) it should be fine.

**Reply by pretendhigh** [reply]
Thanks for your advice and explanation.   It make me belive that it is ok to delete items in dag_run, task_instance and job tables which's end_date is 30 days ago. Of cource, when delete items in dag_run table, I should care about some dag rerun.
